analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -61,9 +62,106 @@ protected override ProgramState PreProcess(ProgramState state, IBinaryOperationW
         _ => null
     };
 
+    private static NumberConstraint CalculateMultiply(NumberConstraint left, NumberConstraint right)
+    {
+        var products = new[] { left.Min * right.Min, left.Min * right.Max, left.Max * right.Min, left.Max * right.Max };
+        var min = products.Min();
+        var max = products.Max();
+
+        if ((left.Min is null && right.CanBePositive) || (right.Min is null && left.CanBePositive)
+            || (left.Max is null && right.CanBeNegative) || (right.Max is null && left.CanBeNegative))
+        {
+            min = null;
+        }
+        if ((left.Min is null && right.CanBeNegative) || (right.Min is null && left.CanBeNegative)
+            || (left.Max is null && right.CanBePositive) || (right.Max is null && left.CanBePositive))
+        {
+            max = null;
+        }
+
+        return NumberConstraint.From(min, max);
+    }
+
+    private static NumberConstraint CalculateAnd(NumberConstraint left, NumberConstraint right)
+    {
+        BigInteger? min = 0;
+        if (left.CanBeNegative && right.CanBeNegative)
+        {
+            min = left.Min.HasValue && right.Min.HasValue
+                ? NegativeMagnitude(BigInteger.Min(left.Min ?? 0, right.Min ?? 0))
+                : null;
+        }
+        // If both operands are negative, the result is negative. Otherwise, it is positive.
+        // If operands are both negative or both positive the result cannot be bigger than the smaller of the two.
+        // If the operands have different signs, the result cannot be bigger than the positive one.
+        BigInteger? max = null;
+        if (left.IsFullyNegative && right.CanBePositive)
+        {
+            max = right.Max;
+        }
+        else if (left.CanBePositive && right.IsFullyNegative)
+        {
+            max = left.Max;
+        }
+        else if (left.IsFullyPositive && right.CanBePositive && right.CanBeNegative)
+        {
+            max = left.Max;
+        }
+        else if (left.CanBePositive && left.CanBeNegative && right.IsFullyPositive)
+        {
+            max = right.Max;
+        }
+        else if (left.CanBePositive && left.CanBeNegative && right.CanBePositive && right.CanBeNegative && left.Max.HasValue && right.Max.HasValue)
+        {
+            max = BigInteger.Max(left.Max.Value, right.Max.Value);
+        }
+        else if ((left.IsFullyNegative && right.IsFullyNegative) || (left.IsFullyPositive && right.IsFullyPositive))
+        {
+            max = SmallestMaximum(left, right);
+        }
+
+        return NumberConstraint.From(min, max);
+    }
+
+    private static BigInteger? NegativeMagnitude(BigInteger value)
+    {
+        // For various inputs, we're looking for the longest chain of 1 from the MSB side
+        // 11000000 - 64
+        // 11000001 - 63
+        // 11000010 - 62
+        // ------------ -
+        // 11000000 - 64 expected result
+        // ^^
+        BigInteger m = -1;
```

Comment:
Not sure why it's `m`. I'd name it `ret` to indicate that it's the returned value
```suggestion
        BigInteger ret = -1;
```